File: code/ex76.cpp
// Sliding window

class Solution {
public:
    string minWindow(string s, string t) {
        int n = s.length();
        
        // Store the cnt of required chars in t
        unordered_map<char, int> cnt;
        for (char c : t)    cnt[c]++;
        
        string res = s + " ";
        int lo = 0, MIN = s.length();
        unordered_map<char, int> cntWindow;    // Store the cnt of required chars in window
        unordered_set<char> inWindow;    // Char required which meets the cnt in window
        for (int hi = 0; hi < n; hi++) {
            // Move hi 1 place to the right
            if (cnt.find(s[hi]) != cnt.end()) {
                if (++cntWindow[s[hi]] >= cnt[s[hi]])
                    inWindow.insert(s[hi]);
            }
            
            // Try to move lo pointer as right as possible
            while (inWindow.size() == cnt.size() || hi - lo + 1 > res.length()) {
                if (inWindow.size() == cnt.size() && hi - lo + 1 < res.length())
                    res = s.substr(lo, hi - lo + 1);
                if (cnt.find(s[lo]) != cnt.end()) {
                    if (--cntWindow[s[lo]] < cnt[s[lo]])
                        inWindow.erase(s[lo]);
                }
                lo++;
            }
        }
        
        return res == s + " " ? "" : res;
    }
};

File: code/ex76.java
class Solution {
    public String minWindow(String s, String t) {
        int n = s.length();
        
        // Get the count of the required characters
        int[] cnt = new int[126];
        for (char c : t.toCharArray())
            cnt[c]++;
        int num = 0;
        for (int i = 0; i < 126; i++)
            if (cnt[i] != 0)    num++;
        
        // Sliding window to find the minimum window including these characters
        int[] cntWindow = new int[126];
        Set<Character> inWindow = new HashSet();
        String res = s + " ";
        int lo = 0;
        for (int hi = 0; hi < n; hi++) {
            // Add character at hi pointer
            char chi = s.charAt(hi);
            if (cnt[chi] != 0) {
                if (++cntWindow[chi] >= cnt[chi])
                    inWindow.add(chi);
            }
            
            // Try to move lo pointer to the right
            while (inWindow.size() == num || hi - lo + 1 > res.length()) {
                if (inWindow.size() == num && s.substring(lo, hi + 1).length() < res.length())
                    res = s.substring(lo, hi + 1);
                char clo = s.charAt(lo);
                if (cnt[clo] != 0) {
                    if (--cntWindow[clo] < cnt[clo])
                        inWindow.remove(clo);
                }
                lo++;
            }
        }
        
        return res.equals(s + " ") ? "" : res;
    }
}